### PR TITLE
RSPEC-3052 : Fields should not be initialized to default values

### PR DESCRIPTION
**Pull request by axel3rd**
Proposal for [RSPEC-3052](http://jira.sonarsource.com/browse/RSPEC-3052). Clean implementation of previous [PR](https://github.com/SonarSource/sonar-java/pull/288).

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/FieldsShouldNotBeInitializedToDefaultValuesCheck.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    try {
+      return initTree.symbolType().isNumerical() && Double.doubleToRawLongBits(NumberFormat.getInstance(Locale.US).parse(initTree.value()).doubleValue()) == 0;
+    } catch (ParseException e) {
+      // Special case when char initialized with '\x' value. This is a numeric primitive but NumberFormat.parse throw exception
+      return initTree.symbolType().isPrimitive(Primitives.CHAR) && "'\\0'".equals(initTree.value());
+    }
+  }
+
+  private boolean isInitBooleanFalse(LiteralTree initTree) {
+    if (initTree.symbolType().isPrimitive(Primitives.BOOLEAN) && !Boolean.parseBoolean(initTree.value())) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/FieldsShouldNotBeInitializedToDefaultValuesCheck.java]: I am pretty sure this if(cond) true else false can be refactored ;)

**Comment by axel3rd** [comment]
@benzonico : Feedbacks done. _The perfect merge will be (perhaps) for next time ^^._

File: java-checks/src/main/java/org/sonar/java/checks/FieldsShouldNotBeInitializedToDefaultValuesCheck.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    if (!isLiteral(fieldTree.initializer())) {
+      return;
+    }
+    LiteralTree initTree = (LiteralTree) fieldTree.initializer();
+    if (isInitObjectNull(initTree) || isInitNumericalZero(initTree) || isInitBooleanFalse(initTree)) {
+      addIssue(fieldTree, String.format("Remove this initialization to \"%s\", the compiler will do that for you.", fieldTree.symbol().name()));
+    }
+  }
+
+  private boolean isLiteral(ExpressionTree exprTree) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/FieldsShouldNotBeInitializedToDefaultValuesCheck.java]: Missing @Nullable annotation here.

**Comment by axel3rd** [comment]
@benzonico : *@Nullable* added.

**Comment by axel3rd** [comment]
@benzonico : I have deployed this PR temporarily in our SonarQube ecosystem (inside a company plugin).

The current implementation raises an issue on :
``
public static final int VAR = 0;
``

But the java compiler could be disagree with (*The blank final field VAR may not have been initialized*) :
``
public static final int VAR;
``

=> The ``final`` fields must be out of the scope of this issue ? Documentation must be updated (This case is not excplicit in [RSPEC-3052](http://jira.sonarsource.com/browse/RSPEC-3052)) ? 
(After check, out of scope of some other implementation : [checkstyle](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ExplicitInitializationCheck.java#L136))

If yes I can come back shortly with a better implementation. Thanks.

**Comment by axel3rd** [comment]
PR update with previous comments.

**Comment by axel3rd** [comment]
[DefaultInitializedFieldCheck](https://github.com/SonarSource/sonar-java/blob/master/java-checks/src/main/java/org/sonar/java/checks/DefaultInitializedFieldCheck.java) implements this PR, in v3.6 ([SONARJAVA-1242](https://jira.sonarsource.com/browse/SONARJAVA-1242)) => closing

**Pull request closed without merging**